test/functorch/test_vmap.py

```diff
@@ -4003,6 +4024,13 @@ def test():
                     continue
                 kwargs = sample_input.kwargs
                 is_batch_norm_and_training = is_batch_norm_training(op.name, kwargs)
+                out_dim = 0
+                if op.name.endswith("CustomOp"):
+                    out_dim_example = get_vmap_out_dims_example_custom_op(op.name)
```

Comment:
Sorry I misunderstood you over chat -- if an operator returns a non-Tensor then we should not be able to vmap it. In that case, we should prevent the user from registering a vmap rule (and we can add a default vmap rule that errors out saying why)

Reply:
Do we not support returning non-Tensor in general or are there some special cases? In general it seems that the default vmap supports returning non-Tensor outputs?

```python
def f(x):
    return x, 1

x = torch.randn(3)
print(torch.vmap(f, out_dims=(0, None))(x))
# (tensor([-0.6958, -0.3569, -0.3476]), 1)
```

Reply:
hmm

Reply:
I think the answer is that it depends on the operator.
- `aten::item(Tensor) -> Scalar` takes a scalar Tensor and returns a Python Number for the element inside it. It doesn't make sense for this to have a batching rule, because if the values in the Tensor are different then there is no way to return a "batched Scalar"
- `aten::numel(Tensor) -> int` that returns the number of elements inside a Tensor. We can have a batching rule for this: it just returns the numel ignoring the batch dimension of the Tensor.

My conclusions on what we should do are:
- we should allow people to write batching rules for all operators. It's up to the user to decide if it makes sense to have a batching rule or not.
- numpy_split_copy_with_int: it makes sense to write a batching rule for this
- we should keep your changes to the test_vmap infrastructure (though I'll leave some more comments around those) so we can test those operators.

How does that sound?

Reply:
Sounds good! Thanks!